dev-db/clickhouse/clickhouse-9999.ebuild

```diff
@@ -0,0 +1,724 @@
+	eapply_user
+	cmake-utils_src_prepare
+
+	local cmd lib
+	if use llvm; then
+		ln --relative --symbolic dbms/programs/clang/Compiler-7.{0.0,1.0} || \
+			die "Failed to fixup symbolic links for internal compiler"
+	fi
+	if use hdfs3; then
+		sed --in-place 's/-D_GLIBCXX_USE_CXX11_ABI=0//' contrib/libhdfs3-cmake/CMakeLists.txt || \
+			die "Failed to fixup libhdfs3"
+	fi
+	if use jemalloc || use tcmalloc; then
+		local component
+		lib="\${$(
+			usex jemalloc JEMALLOC_LIBRARIES "$(
+				usex debug GPERFTOOLS_TCMALLOC_MINIMAL_DEBUG GPERFTOOLS_TCMALLOC_MINIMAL
+			)"
+		)}"
+		for component in dbms clickhouse_common_io; do
+			echo "target_link_libraries(${component} PRIVATE ${lib})" >> dbms/CMakeLists.txt
+		done
+
+		echo "target_link_libraries(clickhouse-odbc-bridge PRIVATE ${lib})" >> \
+			dbms/programs/odbc-bridge/CMakeLists.txt
+		echo "target_link_libraries(zookeeper-adjust-block-numbers-to-parts PRIVATE ${lib})" >> \
+			utils/zookeeper-adjust-block-numbers-to-parts/CMakeLists.txt
+		echo "target_link_libraries(convert-month-partitioned-parts PRIVATE ${lib})" >> \
+			utils/convert-month-partitioned-parts/CMakeLists.txt
+	fi
+	if use kafka; then
+		echo "target_link_libraries(cppkafka PRIVATE \${ZLIB_LIBRARIES})" >> \
+			contrib/cppkafka-cmake/CMakeLists.txt
+	fi
+	if use libunwind; then
+		cmd='s/add_library(unwind_static \${LIBUNWIND_SOURCES}'
+		cmd+='/add_library(unwind_static STATIC \${LIBUNWIND_SOURCES}/'
+
+		if use '!static' && use static-libs; then
+			cmd+=';s/target_compile_options(unwind_static PRIVATE'
+			cmd+='/target_compile_options(unwind_static PRIVATE -fPIC/'
+		fi
+
+		sed --in-place "${cmd}" contrib/libunwind-cmake/CMakeLists.txt || die "Failed to fixup libunwind"
+	fi
+
+	if use '!static' && use static-libs; then
+		for lib in Util Foundation; do
+			echo "target_compile_options(\${LIBNAME} PRIVATE -fPIC)" >> \
+				"contrib/poco/${lib}/CMakeLists.txt" || \
+				die "Failed to add PIC to ${lib}"
+		done
+	fi
+
+	[[ "${PV}" == "9999" ]] && mkdir --parents contrib/boost/boost
+
+	local cmake_config_checks line cmake_var use_flag
+
+	cmake_config_checks="${CLICKHOUSE_CMAKE_CONFIG_CHECKS[@]}"
+
+	cmake_config_checks+=(
+		"CMAKE_INTERPROCEDURAL_OPTIMIZATION $(usex lto)"
+		"USE_RAPIDJSON \
+			$(usex json "$(usex '!cpu_flags_x86_avx2' json '!cpu_flags_x86_avx2')" json)"
+		"USE_SIMDJSON \
+			$(usex json "$(usex cpu_flags_x86_avx2 json cpu_flags_x86_avx2)" json)"
+	)
+
+	for line in "${CLICKHOUSE_CMAKE_CONFIG_CHECKS[@]}"; do
+		read -r cmake_var use_flag <<< "${line}"
+		echo "if ($(
+			[[ "${use_flag}" ]] && usex "${use_flag}" "NOT" "" || echo NOT
+		) ${cmake_var})" >> CMakeLists.txt
+		echo "    message (FATAL_ERROR \"Failed to $(
+			[[ "${use_flag}" ]] && usex "${use_flag}" use disable || echo check
+		) ${use_flag}: ${cmake_var}\")" >> CMakeLists.txt
+		echo "endif ()" >> CMakeLists.txt
+	done
+}
+
+src_configure() {
+	use cpu_flags_x86_sse4_1 && append-cflags "-msse4.1" && append-cxxflags "-msse4.1"
```

Comment:
CPU_FLAGS* are not used to alter CFLAGS.

Reply:
Well, this is important. ClickHouse is performance oriented DB and they make use of CPU features.
One thing they do in order to achieve that is that they test CPU at configure time *and* add `HAVE_***` defines and `-m***` CFLAGS.
Of course, the flags are going to be detected are those that are suitable for host system, so I disabled CPU check. But as a result, no CFLAGS or HAVE_** are going to be set, so I do that manually.

Reply:
In Gentoo people set correct `CFLAGS` that enable CPU features they want.

Reply:
I'm aware of what Gentoo users like to do with CFLAGS. And they are welcomed to.
I'm either setting up this flags to reflect `cpu_flags_x86_**` or parsing current CFLAGS to figure out which `HAVE_***` are needed.
Adding this flags will not hurt. Not adding them can. Also this way binpkg-multi-instance will work properly for CFLAGS provided by official `stage3` or similar to those.

Reply:
If there is any QA check about that, please point me to the corresponding manual and I will disable these checks for that ebuild by setting up variables needed and providing summary of this thread as a comment

Reply:
https://devmanual.gentoo.org/general-concepts/use-flags/index.html that big section about not using USE flags.

Reply:
In clickhouse, it if build system autodetect the CPU Instruction it will add msse4.1 flag, however as in this PR it is disabled, it's now on anyone who builds it to add it.
Not having this option passed will result in much worse user experience. So if you are against making an exception for it, I would vote on removing this ebuild from the tree. It would be better that way.

Reply:
Disable any configure tampering with CFLAGS, so that user's `-march` will work as expected.

Reply:
I'm on the way to provide compromise: I'll issue a warning if correct -m flag is absent. I wouldn't check march/mtune though because bringing users' attention would be enough imo

Reply:
Excuse me but are you serious? The whole purpose of `-march` is to let the compiler enable appropriate instruction sets based on user's CPU without asking him to explicitly figure out what to enable.

Reply:
It is either that or add-flags

Reply:
No, `-march` is the way to go. It is complete nonsense to tell people to pass `-msse4.1` or whatever when it is already implied by `-march`.

Reply:
HAVE_*** defines and CPU features enabled with CFLAGS should be in sync.

Reply:
... and I can't easily figure out all the implicit `-m` flags, so I'm either forcing this flags (which is safe and reasonable) or I `ewarm` that I can't check that thing automatically (protecting users' sacred right to shoot themselves in the foot)

Reply:
I've pushed and update with warnings instead of flag manipulations. I also tracked down one additional place with `-m` flags manipulated and sed-patched it.
Please, mark as resolved or suggest a better alternative that would satisfy both Gentoo QA and ClickHouse devs and users